From f8cf40a2d26b021d87431727101a64fcc8c11ee6 Mon Sep 17 00:00:00 2001
From: robinz1x <robinx.zhang@intel.com>
Date: Fri, 4 May 2018 09:26:49 +0800
Subject: [PATCH 4/5] Adjust sendObject parameter's data type

Now the MtpDevice#sendObject is using int to define object size,
if the object size is bigger than 2GB, the size will be type cast to
a negative number in MtpDevice#sendObject. And it will cause a endless
loop in MtpDataPacket#write until DUT auto reboot.

Due to the JNI api android_mtp_MtpDevice_send_object is using uint32_t
to define object size, so adjust the native api MtpDevice#sendObject
size parameter's data type to uint32_t.

Change-Id: I6d9491713adfdefa750642d792da67dd3e6a9e84
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-64420
Signed-off-by: robinz1x <robinx.zhang@intel.com>
---
 media/mtp/MtpDevice.cpp | 4 ++--
 media/mtp/MtpDevice.h   | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/media/mtp/MtpDevice.cpp b/media/mtp/MtpDevice.cpp
index 0bf7854..89a2460 100644
--- a/media/mtp/MtpDevice.cpp
+++ b/media/mtp/MtpDevice.cpp
@@ -516,7 +516,7 @@ MtpObjectHandle MtpDevice::sendObjectInfo(MtpObjectInfo* info) {
     return (MtpObjectHandle)-1;
 }
 
-bool MtpDevice::sendObject(MtpObjectHandle handle, int size, int srcFD) {
+bool MtpDevice::sendObject(MtpObjectHandle handle, uint32_t size, int srcFD) {
     Mutex::Autolock autoLock(mMutex);
 
     if (mLastSendObjectInfoTransactionID + 1 != mTransactionID ||
@@ -529,7 +529,7 @@ bool MtpDevice::sendObject(MtpObjectHandle handle, int size, int srcFD) {
     if (sendRequest(MTP_OPERATION_SEND_OBJECT)) {
         mData.setOperationCode(mRequest.getOperationCode());
         mData.setTransactionID(mRequest.getTransactionID());
-        const int writeResult = mData.write(mRequestOut, mPacketDivisionMode, srcFD, size);
+        const uint32_t writeResult = mData.write(mRequestOut, mPacketDivisionMode, srcFD, size);
         const MtpResponseCode ret = readResponse();
         return ret == MTP_RESPONSE_OK && writeResult > 0;
     }
diff --git a/media/mtp/MtpDevice.h b/media/mtp/MtpDevice.h
index a9a3e0e..dc97b1d 100644
--- a/media/mtp/MtpDevice.h
+++ b/media/mtp/MtpDevice.h
@@ -104,7 +104,7 @@ public:
     MtpObjectInfo*          getObjectInfo(MtpObjectHandle handle);
     void*                   getThumbnail(MtpObjectHandle handle, int& outLength);
     MtpObjectHandle         sendObjectInfo(MtpObjectInfo* info);
-    bool                    sendObject(MtpObjectHandle handle, int size, int srcFD);
+    bool                    sendObject(MtpObjectHandle handle, uint32_t size, int srcFD);
     bool                    deleteObject(MtpObjectHandle handle);
     MtpObjectHandle         getParent(MtpObjectHandle handle);
     MtpStorageID            getStorageID(MtpObjectHandle handle);
-- 
1.9.1

